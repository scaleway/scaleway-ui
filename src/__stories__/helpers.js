export const showAlert = (text = 'fired!') => {
  // eslint-disable-next-line no-alert
  alert(text)
}

export const log = (text = 'fired!') => {
  // eslint-disable-next-line no-console
  console.log(text)
}
