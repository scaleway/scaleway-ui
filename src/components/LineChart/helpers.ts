import { Theme } from '@emotion/react'
import { DatumValue, Serie } from '@nivo/line'

const parse = (data?: DatumValue | null) => {
  if (typeof data === 'number') return data
  if (typeof data === 'string') return parseFloat(data)
  if (data instanceof Date) return data.getTime()

  return 0
}

export const getMaxChartValue = (preppedData?: Serie[]): number => {
  if (!preppedData?.length) return 0

  const maximum = Math.max(
    ...preppedData.map(data =>
      data.data.reduce(
        (max, { y }) => Math.max(parse(y), max),
        Number.NEGATIVE_INFINITY,
      ),
    ),
  )

  return Math.ceil(maximum + maximum * 0.1)
}

export const getMinChartValue = (preppedData?: Serie[]): number => {
  if (!preppedData?.length) return 0

  const minimum = Math.min(
    ...preppedData.map(data =>
      data.data.reduce(
        (min, { y }) => Math.min(parse(y), min),
        Number.POSITIVE_INFINITY,
      ),
    ),
  )

  return Math.floor(minimum - minimum * 0.1)
}

export const getMin = (values: number[] = []): number =>
  values.length ? Math.min(...values) : 0

export const getMax = (values: number[] = []): number =>
  values.length ? Math.max(...values) : 0

export const getAverage = (values: number[] = []): number =>
  values.length
    ? Math.round(
        (values.reduce((average, curr) => average + curr, 0) / values.length) *
          100,
      ) / 100
    : 0

export const getCurrent = (values: number[] = []): number =>
  values.length ? values[values.length - 1] : 0

export const getSelected = (
  label: string,
  index: number,
  selected: string[] = [],
): string[] => {
  const labelIndexed = label + index.toString()
  const found = selected.indexOf(labelIndexed)

  if (found > -1) {
    selected.splice(found, 1)
  } else {
    selected.push(labelIndexed)
  }

  return selected
}

export const legendColors = (theme: Theme): string[] =>
  [
    theme.colors.chartGreen,
    theme.colors.chartPurple,
    theme.colors.gray350,
  ] as string[]

export const getLegendColor = (index: number, theme: Theme): string => {
  const colors = legendColors(theme)

  return colors[index] || colors[colors.length - 1]
}
