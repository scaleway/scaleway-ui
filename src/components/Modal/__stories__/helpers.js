export const animations = [
  'fold',
  'zoom',
  'scaleUp',
  'scaleBack',
  'sketch',
  'slideTop',
  'slideBottom',
  'slideLeft',
  'slideRight',
]

export const widths = ['xxsmall', 'xsmall', 'small', 'medium', 'large']

export const placements = [
  'center',
  'top-left',
  'top',
  'top-right',
  'right',
  'bottom-right',
  'bottom',
  'bottom-left',
  'left',
]
