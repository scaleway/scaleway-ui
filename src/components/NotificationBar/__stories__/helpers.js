export default [
  'security',
  'info',
  'success',
  'warning',
  'warning-blue',
  'warning-orange',
  'alert-orange',
  'alert-red',
]
