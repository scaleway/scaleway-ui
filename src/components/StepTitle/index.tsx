import styled from '@emotion/styled'
import PropTypes from 'prop-types'
import React, { ReactNode } from 'react'

export const Steps = styled.ul`
  list-style: none;
  padding-left: 16px;
  margin-top: 24px;
  text-align: left;
  font-size: 16px;
`

export const Step = styled('li', {
  shouldForwardProp: prop => !['disabled'].includes(prop.toString()),
})<{ disabled: boolean }>`
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  justify-content: center;
  color: ${({ theme, disabled }) =>
    disabled ? theme.colors.neutral.textWeak : theme.colors.neutral.textStrong};
`

export const Chip = styled('span', {
  shouldForwardProp: prop => !['size'].includes(prop.toString()),
})<{ size: Sizes }>`
  display: flex;
  align-items: center;
  justify-content: center;
  min-width: 32px;
  height: 32px;
  border-radius: 50%;
  border: 1px solid ${({ theme }) => theme.colors.neutral.borderWeak};
  background-color: ${({ theme }) => theme.colors.neutral.background};
  margin-right: ${({ theme, size }) =>
    size === 'medium' ? theme.space['1'] : theme.space['2']};
  text-align: center;
  font-weight: 600;
`

const StyledDiv = styled.div`
  width: 100%;
`

export type Sizes = 'medium' | 'large'

interface StepTitleProps {
  index: number
  size?: Sizes
  disabled?: boolean
  keyPrefix?: string
  children: ReactNode
}

const StepTitle = ({
  index,
  children,
  keyPrefix,
  size = 'medium',
  disabled = false,
}: StepTitleProps) => (
  <Step
    key={`${keyPrefix ? `${keyPrefix}-` : ''}step-${index}`}
    disabled={disabled}
  >
    <Chip size={size}>{`${index}`}</Chip>
    <StyledDiv>{children}</StyledDiv>
  </Step>
)

StepTitle.propTypes = {
  children: PropTypes.node.isRequired,
  disabled: PropTypes.bool,
  /**
   * Add a prefix if you have multiple StepTitle in a same page.
   */
  index: PropTypes.number.isRequired,
  keyPrefix: PropTypes.string,
  size: PropTypes.string,
}

export default StepTitle
