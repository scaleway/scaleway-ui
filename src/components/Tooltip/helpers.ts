import { RefObject } from 'react'

export type TooltipPlacement = 'top' | 'right' | 'bottom' | 'left' | 'auto'
export const ARROW_WIDTH = 6 // in px
export const SPACE = 6 // in px
export const TOTAL_USED_SPACE = ARROW_WIDTH + SPACE // in px
export const DEFAULT_POSITIONS = {
  arrowLeft: 50,
  arrowTop: 99,
  arrowTransform: 'translate(-50%, -50)',
  left: 0,
  rotate: 135,
  tooltipInitialPosition: 'translate3d(0, 0, 0))',
  top: 0,
}

export const TOOLTIP_INITIAL_POSITION = {
  auto: `translate3d(0, ${TOTAL_USED_SPACE}px, 0)`,
  bottom: `translate3d(0, -${TOTAL_USED_SPACE}px, 0)`,
  left: `translate3d(${TOTAL_USED_SPACE}px, 0, 0)`,
  right: `translate3d(-${TOTAL_USED_SPACE}px, 0, 0)`,
  top: `translate3d(0, ${TOTAL_USED_SPACE}px, 0)`,
}

type ComputePlacementTypes = {
  childrenStructuredRef: DOMRect
  tooltipStructuredRef: DOMRect
}

/**
 * This function will find the best placement in a window for tooltip based on children position and tooltip size
 */
export const computePlacement = ({
  childrenStructuredRef,
  tooltipStructuredRef,
}: ComputePlacementTypes) => {
  const {
    top: childrenTop,
    left: childrenLeft,
    right: childrenRight,
  } = childrenStructuredRef

  const { width: tooltipWidth, height: tooltipHeight } = tooltipStructuredRef

  if (childrenTop - tooltipHeight - TOTAL_USED_SPACE < 0) {
    return 'bottom'
  }

  if (childrenLeft - tooltipWidth - TOTAL_USED_SPACE < 0) {
    return 'right'
  }

  if (childrenRight + tooltipWidth + TOTAL_USED_SPACE > window.innerWidth) {
    return 'left'
  }

  return 'top'
}

type ComputePositionsTypes = {
  placement: TooltipPlacement
  childrenRef: RefObject<HTMLDivElement>
  tooltipRef: RefObject<HTMLDivElement>
}

/**
 * This function will compute the positions of tooltip and arrow based on children position and tooltip size
 */
export const computePositions = ({
  placement,
  childrenRef,
  tooltipRef,
}: ComputePositionsTypes) => {
  const childrenStructuredRef = (
    childrenRef.current as HTMLDivElement
  ).getBoundingClientRect()
  const tooltipStructuredRef = (
    tooltipRef.current as HTMLDivElement
  ).getBoundingClientRect()

  const placementBasedOnWindowSize =
    placement === 'auto'
      ? computePlacement({
          childrenStructuredRef,
          tooltipStructuredRef,
        })
      : placement

  const {
    top: childrenTop,
    left: childrenLeft,
    right: childrenRight,
    width: childrenWidth,
    height: childrenHeight,
  } = childrenStructuredRef

  const { width: tooltipWidth, height: tooltipHeight } = tooltipStructuredRef

  switch (placementBasedOnWindowSize) {
    case 'bottom':
      return {
        arrowLeft: tooltipWidth / 2,
        arrowTop: -ARROW_WIDTH - 5,
        arrowTransform: '',
        left: childrenLeft + childrenWidth / 2 - tooltipWidth / 2,
        rotate: 180,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.bottom,
        top: childrenTop + childrenHeight + ARROW_WIDTH + SPACE,
      }
    case 'left':
      return {
        arrowLeft: tooltipWidth + ARROW_WIDTH + 5,
        arrowTop: tooltipHeight / 2,
        arrowTransform: 'translate(-50%, -50%)',
        left: childrenLeft - tooltipWidth - ARROW_WIDTH - SPACE * 2,
        rotate: -90,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.left,
        top: childrenTop - tooltipHeight / 2 + childrenHeight / 2,
      }
    case 'right':
      return {
        arrowLeft: -ARROW_WIDTH - 5,
        arrowTop: tooltipHeight / 2,
        arrowTransform: 'translate(50%, -50%)',
        left: childrenRight + ARROW_WIDTH + SPACE * 2,
        rotate: 90,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.right,
        top: childrenTop - tooltipHeight / 2 + childrenHeight / 2,
      }
    default: // top placement is default value
      return {
        arrowLeft: tooltipWidth / 2,
        arrowTop: tooltipHeight - 1,
        arrowTransform: '',
        left: childrenLeft + childrenWidth / 2 - tooltipWidth / 2,
        rotate: 0,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.top,
        top: childrenTop - tooltipHeight - ARROW_WIDTH - SPACE,
      }
  }
}
