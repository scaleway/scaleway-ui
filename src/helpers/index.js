export * from './clipboard'

export const parseIntOr = (str, fallback) => {
  const value = Number.parseInt(str, 10)
  return !Number.isNaN(value) ? value : fallback
}

export const onKeyOnlyNumbers = e => {
  // `e.which` and `e.keyCode` are deprecated (https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent)
  const keyCode = e.key.charCodeAt(0)
  if (keyCode < 48 || keyCode > 57) {
    e.preventDefault()
  }
}
