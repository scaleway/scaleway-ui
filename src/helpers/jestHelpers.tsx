/* eslint-disable import/no-extraneous-dependencies */
import { ThemeProvider } from '@emotion/react'
import makeHelpers from '@scaleway/jest-helpers'
import React, { FC } from 'react'
import defaultTheme from '../theme'

interface WrapperProps {
  theme?: typeof defaultTheme
}

const Wrapper: FC<WrapperProps> = ({ theme = defaultTheme, children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotWithPortal,
} = makeHelpers(Wrapper, { classNameReplacer: className => className })
