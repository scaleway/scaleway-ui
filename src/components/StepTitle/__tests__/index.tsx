import React from 'react'
import StepTitle from '..'
import { shouldMatchEmotionSnapshot } from '../../../helpers/jestHelpers'

describe('StepTitle', () => {
  test('renders correctly ', () =>
    shouldMatchEmotionSnapshot(<StepTitle index={1}>Item 1</StepTitle>))

  test('renders correctly with custom keyPrefix', () =>
    shouldMatchEmotionSnapshot(
      <StepTitle index={1} keyPrefix="test">
        Item 1
      </StepTitle>,
    ))

  test('renders correctly with large size', () =>
    shouldMatchEmotionSnapshot(
      <StepTitle index={1} size="large">
        Item 1 large
      </StepTitle>,
    ))

  test('renders correctly with disabled state', () =>
    shouldMatchEmotionSnapshot(
      <StepTitle index={1} disabled>
        <div>Item 1 with disabled state</div>
      </StepTitle>,
    ))
})
