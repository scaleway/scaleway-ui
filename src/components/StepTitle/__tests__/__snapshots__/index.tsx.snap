// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulletList renders correctly  1`] = `
<DocumentFragment>
  .cache-7078ng-Steps {
  list-style: none;
  padding-left: 16px;
  margin-top: 24px;
  text-align: left;
  font-size: 16px;
}

.cache-z01g5q-Step {
  margin-bottom: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.cache-1da3zhe-Chip {
  display: inline-block;
  min-width: 24px;
  height: 24px;
  border-radius: 12px;
  background-color: #4a4f62;
  color: #ffffff;
  margin-right: 8px;
  text-align: center;
  font-weight: 600;
}

.cache-1talnks-StyledDiv {
  width: 100%;
}

<ul
    class="cache-7078ng-Steps e12g85273"
  >
    <li
      class="cache-z01g5q-Step e12g85272"
    >
      <span
        class="cache-1da3zhe-Chip e12g85271"
      >
        1
      </span>
      <div
        class="cache-1talnks-StyledDiv e12g85270"
      >
        <div>
          a
        </div>
      </div>
    </li>
    <li
      class="cache-z01g5q-Step e12g85272"
    >
      <span
        class="cache-1da3zhe-Chip e12g85271"
      >
        2
      </span>
      <div
        class="cache-1talnks-StyledDiv e12g85270"
      >
        <div>
          b
        </div>
      </div>
    </li>
    <li
      class="cache-z01g5q-Step e12g85272"
    >
      <span
        class="cache-1da3zhe-Chip e12g85271"
      >
        3
      </span>
      <div
        class="cache-1talnks-StyledDiv e12g85270"
      >
        <div>
          c
        </div>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`BulletList renders correctly with custom keyPrefix 1`] = `
<DocumentFragment>
  .cache-7078ng-Steps {
  list-style: none;
  padding-left: 16px;
  margin-top: 24px;
  text-align: left;
  font-size: 16px;
}

.cache-z01g5q-Step {
  margin-bottom: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.cache-1da3zhe-Chip {
  display: inline-block;
  min-width: 24px;
  height: 24px;
  border-radius: 12px;
  background-color: #4a4f62;
  color: #ffffff;
  margin-right: 8px;
  text-align: center;
  font-weight: 600;
}

.cache-1talnks-StyledDiv {
  width: 100%;
}

<ul
    class="cache-7078ng-Steps e12g85273"
  >
    <li
      class="cache-z01g5q-Step e12g85272"
    >
      <span
        class="cache-1da3zhe-Chip e12g85271"
      >
        1
      </span>
      <div
        class="cache-1talnks-StyledDiv e12g85270"
      >
        <div>
          a
        </div>
      </div>
    </li>
    <li
      class="cache-z01g5q-Step e12g85272"
    >
      <span
        class="cache-1da3zhe-Chip e12g85271"
      >
        2
      </span>
      <div
        class="cache-1talnks-StyledDiv e12g85270"
      >
        <div>
          b
        </div>
      </div>
    </li>
    <li
      class="cache-z01g5q-Step e12g85272"
    >
      <span
        class="cache-1da3zhe-Chip e12g85271"
      >
        3
      </span>
      <div
        class="cache-1talnks-StyledDiv e12g85270"
      >
        y
      </div>
    </li>
    <li
      class="cache-z01g5q-Step e12g85272"
    >
      <span
        class="cache-1da3zhe-Chip e12g85271"
      >
        4
      </span>
      <div
        class="cache-1talnks-StyledDiv e12g85270"
      >
        <div>
          c
        </div>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
